galsim/wfirst/wfirst_wcs.py

```diff
@@ -0,0 +1,571 @@
+        # Populate some necessary variables in the FITS header that are always the same, regardless of
+        # input and SCA number.
+        _populate_required_fields(header)
+
+        # And populate some things that just depend on the overall locations or other input, not on
+        # the SCA.
+        header['RA_TARG'] = (use_ra / galsim.degrees, "right ascension of the target (deg) (J2000)")
+        header['DEC_TARG'] = (use_dec / galsim.degrees, "declination of the target (deg) (J2000)")
+        header['PA_OBSY'] = (pa_obsy / galsim.degrees, "position angle of observatory Y axis (deg)")
+        header['PA_FPA'] = (pa_fpa / galsim.degrees, "position angle of FPA Y axis (deg)")
+
+        # Finally do all the SCA-specific stuff.
+        header['SCA_NUM'] = (i_sca, "SCA number (1 - 18)")
+
+        # Set the position of center of this SCA in focal plane angular coordinates.
+        sca_xc_fpa = np.arctan(sca_xc_mm[i_sca]/focal_length)*galsim.radians
+        sca_yc_fpa = np.arctan(sca_yc_mm[i_sca]/focal_length)*galsim.radians
+
+        # Figure out tangent plane positions after distortion, and subtract off those for FPA center
+        # (calculated in header).
+        sca_xc_tp, sca_yc_tp = _det_to_tangplane_positions(sca_xc_fpa, sca_yc_fpa)
+        sca_xc_tp_f = sca_xc_tp - xc_fpa_tp
+        sca_yc_tp_f = sca_yc_tp - yc_fpa_tp
+
+        # Convert to polar coordinates in tangent plane: 
+        # Define theta, phi as in Calabretta & Greisen 2002 A&A 395 1077, fig 3
+        # and eqns 12-15
+        rtheta = np.sqrt(sca_xc_tp_f.rad()**2 + sca_yc_tp_f.rad()**2)*galsim.radians
+        theta = np.arctan(1./rtheta.rad())*galsim.radians
+        # phi is measured clockwise from the -Ytp axis
+        delta_phi = np.arctan2(-sca_xc_tp_f.rad(), sca_yc_tp_f.rad())*galsim.radians
+        delta_phi -= pa_fpa
+        # Leave phi_p at 180 (0 if dec_targ=-90), so that tangent plane axes remain oriented along
+        # celestial coordinates.
+        if use_dec / galsim.degrees > -90.:
+            phi_p = np.pi*galsim.radians
+        else:
+            phi_p = 0.*galsim.radians
+        phi = delta_phi + phi_p
+        cos_delta_phi = np.cos(delta_phi.rad())
+        sin_delta_phi = np.sin(delta_phi.rad())
+        cos_theta = np.cos(theta.rad())
+        sin_theta = np.sin(theta.rad())
+        cos_decp = np.cos(use_dec.rad())
+        sin_decp = np.sin(use_dec.rad())
+
+        # Compute RA, DEC of center of SCA.
+        # (Add pos_targ offsets when implemented.)
+        crval1 = np.arctan2(-sin_delta_phi*cos_theta,
+                             sin_theta*cos_decp - cos_theta*sin_decp*cos_delta_phi)*galsim.radians
+        crval1 += use_ra
+        header['CRVAL1'] = (crval1 / galsim.degrees, "first axis value at reference pixel")
+        crval2 = np.arcsin(sin_theta*sin_decp + cos_theta*cos_decp*cos_delta_phi)*galsim.radians
+        header['CRVAL2'] = (crval2 / galsim.degrees, "second axis value at reference pixel")
```

Comment:
I just pushed a commit that changes the code so that it looks like this:
```
if True:
    ...do things the old way...
    print 'old calculation: ',crval
else:
    ...do things the new way...
    print 'new calculation: ',crval
```

So you can run it for some position and see the old calculation of ```crval```, then change the True to False and run it again to see the new calculation of ```crval```.  You'll see that the old and new agree very precisely for the first SCA then deviate after that.  Like I said, i smell a bug but at this point I am not sure where...